tests/test_backend_general.py

```diff
@@ -0,0 +1,630 @@
+    m.build(backend=backend)
+    m.solve()
+    return m
+
+
+@pytest.fixture(scope="class")
+def infeasible_model_cls(backend) -> calliope.Model:
+    """If we increase demand, the solved model becomes infeasible"""
+    m = build_model({}, "simple_supply,two_hours,investment_costs")
+    m.build(backend=backend)
+    m.backend.update_parameter(
+        "sink_use_equals",
+        m.inputs.sink_use_equals.where(m.inputs.techs == "test_demand_elec") * 100,
+    )
+    with pytest.warns(
+        calliope.exceptions.BackendWarning, match="Model solution was non-optimal"
+    ):
+        m.solve(force=True)
+    return m
+
+
+@pytest.fixture(scope="class")
+def solved_model_cls(backend) -> calliope.Model:
+    m = build_model({}, "simple_supply,two_hours,investment_costs")
+    m.build(backend=backend)
+    m.solve()
+    return m
+
+
+@pytest.fixture()
+def built_model_func_updated_cost_flow_cap(backend, dummy_int: int) -> calliope.Model:
+    m = build_model({}, "simple_supply,two_hours,investment_costs")
+    m.build(backend=backend)
+    m.backend.verbose_strings()
+    m.backend.update_parameter("cost_flow_cap", dummy_int)
+    return m
+
+
+@pytest.fixture(scope="class")
+def solved_model_milp_cls(backend) -> calliope.Model:
+    m = build_model({}, "supply_milp,two_hours,investment_costs")
+    m.build(backend=backend)
+    m.solve()
+    return m
+
+
+class TestBackend:
+    def test_has_backend(self, solved_model_cls):
+        """building a model produces a backend."""
+        assert hasattr(solved_model_cls, "backend")
+
+    def test_to_lp(self, solved_model_cls, tmp_path):
+        """Any built backend should be able to store an LP representation of the math."""
+        filepath = tmp_path / "out.lp"
+        solved_model_cls.backend.to_lp(filepath)
+
+
+class TestOptimality:
+    def test_optimal(self, solved_model_cls):
+        """Solved model is optimal."""
+        assert hasattr(solved_model_cls, "results")
+        assert solved_model_cls._model_data.attrs["termination_condition"] == "optimal"
+
+    def test_solve_non_optimal(self, infeasible_model_cls):
+        """Infeasible models have no results"""
+        assert (
+            infeasible_model_cls._model_data.attrs["termination_condition"]
+            == "infeasible"
+        )
+        assert not infeasible_model_cls.results.data_vars
+
+
+class TestGetters:
+    @pytest.mark.parametrize(
+        "component_type", ["variable", "global_expression", "parameter", "constraint"]
+    )
+    def test_get_missing_component(self, solved_model_cls, component_type):
+        """Try and fail to get a component that doesn't exist in the backend dataset."""
+        with pytest.raises(KeyError):
+            getattr(solved_model_cls.backend, f"get_{component_type}")("foo")
+
+    def test_get_variable(self, solved_model_cls):
+        """Check a decision variable has all expected attributes."""
+        var = solved_model_cls.backend.get_variable("flow_cap")
```

Comment:
Nice-to-have: I'd make this a fixture and test for each requirement separately. Shorter tests are usually easier to maintain.

Ditto for other tests of this type in this file.